<a href="https://colab.research.google.com/github/jihannaurah/BOOTCAMP/blob/main/MINI_PROJECT_BOOTCAMP_DIBIMBING.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
df = pd.read_csv('data.csv', encoding='latin1')
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


DATA QUALITY CHECK

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [3]:
df.isnull().sum()

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0


In [4]:
df.duplicated().sum()

np.int64(5268)

In [5]:
df[df.duplicated()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,12/1/2010 11:45,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,12/1/2010 11:45,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,12/1/2010 11:45,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,12/1/2010 11:45,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,12/1/2010 11:49,2.95,17920.0,United Kingdom
...,...,...,...,...,...,...,...,...
541675,581538,22068,BLACK PIRATE TREASURE CHEST,1,12/9/2011 11:34,0.39,14446.0,United Kingdom
541689,581538,23318,BOX OF 6 MINI VINTAGE CRACKERS,1,12/9/2011 11:34,2.49,14446.0,United Kingdom
541692,581538,22992,REVOLVER WOODEN RULER,1,12/9/2011 11:34,1.95,14446.0,United Kingdom
541699,581538,22694,WICKER STAR,1,12/9/2011 11:34,2.10,14446.0,United Kingdom


DATA CLEANING

In [6]:
df = df.drop_duplicates()

In [7]:
df = df.dropna(subset=['CustomerID'])

In [8]:
print(df.shape)

(401604, 8)


In [9]:
df.isnull().sum()

,0
InvoiceNo,0
StockCode,0
Description,0
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,0
Country,0


In [10]:
df = df.reset_index(drop=True)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 401604 entries, 0 to 401603
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    401604 non-null  object 
 1   StockCode    401604 non-null  object 
 2   Description  401604 non-null  object 
 3   Quantity     401604 non-null  int64  
 4   InvoiceDate  401604 non-null  object 
 5   UnitPrice    401604 non-null  float64
 6   CustomerID   401604 non-null  float64
 7   Country      401604 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 24.5+ MB


FILTERING "QUANTITY, UNITPRICE"

In [12]:
df = df[df['Quantity'] > 0]

In [13]:
df = df[df['UnitPrice'] > 0]

In [14]:
df = df.reset_index(drop=True)

In [15]:
print("Hasil:")
print(df.shape)

Hasil:
(392692, 8)


In [16]:
print("\ninfo detail:")
df.info()


info detail:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 392692 entries, 0 to 392691
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    392692 non-null  object 
 1   StockCode    392692 non-null  object 
 2   Description  392692 non-null  object 
 3   Quantity     392692 non-null  int64  
 4   InvoiceDate  392692 non-null  object 
 5   UnitPrice    392692 non-null  float64
 6   CustomerID   392692 non-null  float64
 7   Country      392692 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 24.0+ MB


MASKING

In [17]:
df['CustomerID'] = df['CustomerID'].astype(str).str[:2] + '***' + df['CustomerID'].astype(str).str[-2:]
df['StockCode'] = df['StockCode'].astype(str).str[:2] + '***' + df['StockCode'].astype(str).str[-2:]
df['InvoiceNo'] = df['InvoiceNo'].astype(str).str[:2] + '***' + df['InvoiceNo'].astype(str).str[-2:]
print("Data Masking")
df.sample(20)

Data Masking


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
233241,56***90,85***1B,RED CHARLIE+LOLA PERSONAL DOORSIGN,5,9/9/2011 13:20,0.39,12***.0,United Kingdom
384077,58***96,22***14,HOT WATER BOTTLE TEA AND SYMPATHY,2,12/5/2011 16:04,4.25,17***.0,United Kingdom
302055,57***26,21***66,COOK WITH WINE METAL SIGN,3,10/25/2011 15:19,2.08,16***.0,United Kingdom
193203,56***53,82***82,AREA PATROLLED METAL SIGN,24,7/27/2011 11:47,2.10,13***.0,United Kingdom
316612,57***40,23***33,IVORY WICKER HEART MEDIUM,6,11/3/2011 12:45,1.25,15***.0,United Kingdom
227623,56***46,23***11,VINTAGE CHRISTMAS STOCKING,6,9/4/2011 15:19,2.55,16***.0,United Kingdom
62989,54***38,22***61,GLASS JAR DAISY FRESH COTTON WOOL,6,2/24/2011 10:34,2.95,12***.0,Saudi Arabia
348467,57***33,22***51,PLACE SETTING WHITE HEART,24,11/18/2011 8:47,0.42,13***.0,United Kingdom
37964,54***31,22***25,ALARM CLOCK BAKELIKE CHOCOLATE,4,1/20/2011 10:47,3.75,12***.0,France
159283,55***29,23***82,TOILET SIGN OCCUPIED OR VACANT,2,6/16/2011 19:59,0.83,17***.0,United Kingdom


**DATA DICTIONARY**
| Nama Kolom | Tipe | Penjelasan Singkat | Contohnya |
| :--- | :--- | :--- | :--- |
| **InvoiceNo** | Teks | Kode transaksi. Kalau ada huruf 'C' di depan, berarti itu barang yang dicancel/retur. | 536365 |
| **StockCode** | Teks | Kode buat tiap barang. (masking). | 85*** |
| **Description** | Teks | Nama barang atau penjelasan produk yang dibeli pelanggan. | BOX OF 6 MINI VINTAGE CRACKERS... |
| **Quantity** | Angka | Jumlah barang yang dibeli. Udah di filter jadi cuma angka yang di atas 0. | 6 |
| **InvoiceDate** | Teks | Tanggal sama jam kapan transaksi itu terjadi. | 12/1/2010 8:26 |
| **UnitPrice** | Desimal | Harga satu barang dalam Poundsterling (£). | 2.55 |
| **CustomerID** | Teks | ID Pelanggan. Udah di sensor (masking) juga buat jaga data sensitif. | 17*** |
| **Country** | Teks | Negara tempat belanja atau asal pelanggannya. | United Kingdom |

In [18]:
df.to_csv('Online_Retail_E-Commerce.csv', index=False)
df.to_excel('Online_Retail_E-Commerce.xlsx', index=False)
print('Tuntas!')

Tuntas!
